irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -51,54 +39,52 @@ namespace iroha {
         log_->error("could not create mutable storage: {}", e->error);
         return;
       }
-      auto storage = std::move(
-          boost::get<
-              expected::Value<std::unique_ptr<ametsuchi::MutableStorage>>>(
-              &mutable_storage_var)
-              ->value);
-
+      auto storage =
+          std::move(
+              boost::get<
+                  expected::Value<std::unique_ptr<ametsuchi::MutableStorage>>>(
+                  mutable_storage_var))
+              .value;
+
+      auto commit = rxcpp::observable<>::just(commit_message);
       SynchronizationEvent result;
 
-      if (validator_->validateBlock(commit_message, *storage)) {
-        storage->apply(*commit_message);
+      if (validator_->validateChain(commit, *storage)) {
         mutable_factory_->commit(std::move(storage));
 
-        result = {rxcpp::observable<>::just(commit_message),
-                  SynchronizationOutcomeType::kCommit};
+        result = {commit, SynchronizationOutcomeType::kCommit};
       } else {
         auto hash = commit_message->hash();
 
-        // while blocks are not loaded and not committed
-        while (storage) {
-          for (const auto &peer_signature : commit_message->signatures()) {
-            auto network_chain = block_loader_->retrieveBlocks(
-                shared_model::crypto::PublicKey(peer_signature.publicKey()));
-
-            std::vector<std::shared_ptr<shared_model::interface::Block>> blocks;
-            network_chain.as_blocking().subscribe(
-                [&blocks](auto block) { blocks.push_back(block); });
-            if (blocks.empty()) {
-              log_->info("Downloaded an empty chain");
-              continue;
-            }
+        [&] {
```

Comment:
Why lambda?

Reply:
To break from a nested loop. I will refactor it as a separate function then.